test/python/transpiler/test_apply_layout.py

```diff
@@ -0,0 +1,124 @@
+from qiskit.transpiler.layout import Layout
+from qiskit.transpiler.passes import ApplyLayout
+from qiskit.transpiler.exceptions import TranspilerError
+
+
+class TestApplyLayout(QiskitTestCase):
+    """ Tests the ApplyLayout pass."""
+
+    def test_trivial(self):
+        """Test if the bell circuit with virtual qubits is transformed into
+        the circuit with physical qubits under trivial layout.
+        """
+        v = QuantumRegister(2, 'v')
+        circuit = QuantumCircuit(v)
+        circuit.h(v[0])
+        circuit.cx(v[0], v[1])
+
+        q = QuantumRegister(2, 'q')
+        expected = QuantumCircuit(q)
+        expected.h(q[0])
+        expected.cx(q[0], q[1])
+
+        dag = circuit_to_dag(circuit)
+        pass_ = ApplyLayout()
+        pass_.property_set['layout'] = Layout({v[0]: 0, v[1]: 1})
+        after = pass_.run(dag)
+
+        self.assertEqual(circuit_to_dag(expected), after)
+
+    def test_raise_when_no_layout_is_supplied(self):
+        """Test error is raised if no layout is found in property_set.
+        """
+        v = QuantumRegister(2, 'v')
+        circuit = QuantumCircuit(v)
+        circuit.h(v[0])
+        circuit.cx(v[0], v[1])
+
+        dag = circuit_to_dag(circuit)
+        pass_ = ApplyLayout()
+        with self.assertRaises(TranspilerError):
+            pass_.run(dag)
+
+    def test_raise_when_no_full_layout_is_given(self):
+        """Test error is raised if no full layout is given.
+        """
+        v = QuantumRegister(2, 'v')
+        circuit = QuantumCircuit(v)
+        circuit.h(v[0])
+        circuit.cx(v[0], v[1])
+
+        dag = circuit_to_dag(circuit)
+        pass_ = ApplyLayout()
+        pass_.property_set['layout'] = Layout({v[0]: 2, v[1]: 1})
+        with self.assertRaises(TranspilerError):
+            pass_.run(dag)
+
+    def test_circuit_with_swap_gate(self):
+        """Test if a virtual circuit with one swap gate is transformed into
+        a circuit with physical qubits.
+
+        [Circuit with virtual qubits]
+          v0:--X---.---M(v0->c0)
+               |   |
+          v1:--X---|---M(v1->c1)
+                   |
+          v2:-----(+)--M(v2->c2)
+
+         Initial layout: {v[0]: 2, v[1]: 1, v[2]: 0}
+
+        [Circuit with physical qubits]
+          q2:--X---.---M(q2->c0)
+               |   |
+          q1:--X---|---M(q1->c1)
+                   |
+          q0:-----(+)--M(q0->c2)
+        """
+        v = QuantumRegister(3, 'v')
+        cr = ClassicalRegister(3, 'c')
+        circuit = QuantumCircuit(v, cr)
+        circuit.swap(v[0], v[1])
+        circuit.cx(v[0], v[2])
+        circuit.measure(v[0], cr[0])
+        circuit.measure(v[1], cr[1])
+        circuit.measure(v[2], cr[2])
+
+        q = QuantumRegister(3, 'q')
+        expected = QuantumCircuit(q, cr)
+        expected.swap(q[2], q[1])
+        expected.cx(q[2], q[0])
+        expected.measure(q[2], cr[0])
+        expected.measure(q[1], cr[1])
+        expected.measure(q[0], cr[2])
```

Comment:
can you change the way you write the expected physical circuits? I think the fact that they are defined over a register named "q" is an artifact of us not being able to support register-less circuits now. Soon, we should just be able to define them over integers. So for this block I would write something like:
```python
        expected = QuantumCircuit(3, 3)
        expected.swap(2, 1)
        expected.cx(2, 0)
        expected.measure(2, 0)
        expected.measure(1, 1)
        expected.measure(0, 2)
```
which is also easier to match with the Layout object, which specifies: `{v[0]: 2, v[1]: 1, v[2]: 0}`

Reply:
This means ApplyLayout also changes/flattens classical registers (to integer addresses), i.e. it is necessary to create new classical register in ApplyLayout like quantum register, right?
I have no concern about quantum register.

Related to this, but may be an another (more general) issue, I think we need the way to restore original "classical" register names (from integer bit addresses). I think people will not care about quantum register name after the circuit is compiled, but still care about classical register name. For example, they prepare ClassicalRegister(1, "answer") for what they want to compute and ClassicalRegister(3, "ancilla") for variables to use in the middle. In current user interface, there is no good way to get the value of answer[0] only. They have to filter the resulting bit string to get the answer bits based on implicit classical-bits layout. That will become difficult when we have much more (qu)bits in the future.

Reply:
Ok sure. My comment might have been a bit premature for now, because we don't really support register-less circuits yet. So I'm ok keeping it as is.

Yes we probably don't want to flatten classical registers, as classical registers can have meaning on the device.